std/format.d

```diff
@@ -8442,35 +8433,1043 @@ printFloat_done:
+    size_t left = max_left;
+    size_t right = max_left;
+
+    // for rounding we need to know if the rest of the number is exactly 0, between 0 and 0.5, 0.5 or above 0.5
+    enum roundType { ZERO, LOWER, FIVE, UPPER }
+    roundType next;
+
+    // Depending on exp, we will use one of three algorithms:
+    //
+    // Algorithm A: For large exponents (exp >= T.mant_dig)
+    // Algorithm B: For small exponents (exp < T.mant_dig - 61)
+    // Algorithm C: For exponents close to 0.
+    //
+    // Algorithm A:
+    //   The number to print looks like this: mantissa followed by several zeros.
+    //
+    //   We know, that there is no fractional part, so we can just use integer division,
+    //   consecutivly dividing by 10 and writing down the remainder from right to left.
+    //   Unfortunately the integer is too large to fit in an ulong, so we use something
+    //   like BigInt: An array of ulongs. We only use 60 bits of that ulongs, because
+    //   this simplifies (and speeds up) the division to come.
+    //
+    //   For the division we use integer division with reminder for each ulong and put
+    //   the reminder of each step in the first 4 bits of ulong of the next step (think of
+    //   long division for the rationale behind this). The final reminder is the next
+    //   digit (from right to left).
+    //
+    // Algorithm B:
+    //   The number to print looks like this: zero dot several zeros followed by the mantissa
+    //
+    //   We know, that the number has no integer part. The algorithm consecutivly multiplies
+    //   by 10. The integer part (rounded down) after the multiplication is the next digit
+    //   (from left to right). This integer part is removed after each step.
+    //   Again, the number is represented as an array of ulongs, with only 60 bits used of
+    //   every ulong.
+    //
+    //   For the multiplication we use normal integer multiplication, which can result in digits
+    //   in the uppermost 4 bits. These 4 digits are the carry which is added to the result
+    //   of the next multiplication and finally the last carry is the next digit.
+    //
+    //   The calculation will stop, when only zeros remain or when we've got enough digits
+    //   for the requested precision. In the second case, we have to find out, which rounding
+    //   we have. Aside from special cases we do this by calculating one more digit.
+    //
+    // Algorithm C:
+    //   This time, we know, that the integral part and the fractional part each fit into a
+    //   ulong. The mantissa might be partially in both parts or completely in the fractional
+    //   part.
+    //
+    //   We first calculate the integral part by consecutive division by 10. Then we calculate
+    //   the fractional part by consecutive multiplication by 10. Again only until we have enough
+    //   digits. Finally, we decide the rounding type, mainly by looking at the next digit.
+
+    ulong[18] bigbuf;
+    if (exp >= T.mant_dig)
```

Comment:
You mention "does not fit into a ulong" below.  Are you assuming that `mant_dig` is always <= 64?  What happens when the mant_dig is 106 or 113?

Reply:
This PR targets only floats, doubles and 64bit reals. See the template constraints of this function. 
The "does not fit" comment refers to the unrolled number, where all the zeros (counted in `exp`) are attached either left or right to the mantissa; these are really large numbers, taking up to a little bit more than 1000 bits.